### PR TITLE
Creation of the CloudSQL utility

### PR DESCRIPTION
**Pull request by TylerTollefson**
For this we have created the base structure for our CloudSQL utility, which creates a sql database instance in the cloud. This will be expanded upon in the future to do table creation, queries, and other sql operations.

### CODE REVIEW

File: src/main/python/utils/cloudsql.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            if "error" in result:
+                raise Exception(result["error"])
+            return result
+
+        time.sleep(1)
+    return operation_complete
+
+@dataclass_json
+@dataclass
+class CreateInstanceConfig():
```
Comment (TylerTollefson) [inline src/main/python/utils/cloudsql.py]: remove all defaults

File: src/main/python/utils/cloudsql.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    enable_ipv4: bool = False
+    # This will require an SSL connection to connect to the db istance
+    require_ssl: bool = False
+    # The private network that is used when allocating the instance a private IP address
+    private_network: str = "projects/som-rit-phi-starr-dev/global/networks/default"
+
+def create_instance(project_id, config: CreateInstanceConfig):
+    credentials = GoogleCredentials.get_application_default()
+    cloudsql = discovery.build('sqladmin', 'v1beta4', credentials=credentials)
+    request_config = {
```
Comment (TylerTollefson) [inline src/main/python/utils/cloudsql.py]: Have this be on the user instead, allow them to pass a json object

File: src/main/python/utils/cloudsql.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    config = Path(args.config).read_text()
+
+    if args.credentials is not None:
+        os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = args.credentials
+
+    if args.project_id is not None:
+        os.environ['GCLOUD_PROJECT'] = args.project_id
+
+    if args.command == "create_instance":
+        create_instance(args.project_id, config=CreateInstanceConfig.from_json(config))
```
Comment (TylerTollefson) [inline src/main/python/utils/cloudsql.py]: just pass config

File: src/main/python/utils/cloudsql.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    operation = cloudsql.instances().insert(project=project_id, body=request_config).execute()
+    if(wait_for_operation(cloudsql, config.project_id, operation["name"])):
+        print("SQL Instance creation finished.")
+
+def main():
+    parser = argparse.ArgumentParser(description="Google CloudSql utility")
+
+    parser.add_argument("--project_id", required=False, help="Your Google Cloud project ID.")
+    parser.add_argument('--credentials', required=False, help='Specify path to a GCP JSON credentials file')
+    parser.add_argument('command', choices=['create_instance'], type=str.lower, help='command to execute')
```
Comment (TylerTollefson) [inline src/main/python/utils/cloudsql.py]: change to insert

File: src/main/wdl/cloudsql.wdl [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        region: { description: "Region in GCP that the instance should be created in" }
+        databaseVersion: { description: "The type of database instance to create" }
+        tier: { description: "Specifications for the machine that the instance will be hosted on" }
+        enableIpv4: { description: "Setting this to true will give the instance a public IP address" }
+        requireSSL: { description: "This will require an SSL connection to connect to the db istance" }
+        privateNetwork: { description: "The private network that is used when allocating the instance a private IP address" }
+    }
+
+    input {
+        String projectId
```
Comment (TylerTollefson) [inline src/main/wdl/cloudsql.wdl:20]: add defaults here

Comment (TylerTollefson) [reply src/main/wdl/cloudsql.wdl:20]: Not specifically project but the other ones

File: src/main/python/utils/cloudsql.py (hunk 1 above)
Comment (TylerTollefson) [reply src/main/python/utils/cloudsql.py]: fixed

File: src/main/python/utils/cloudsql.py (hunk 2 above)
Comment (TylerTollefson) [reply src/main/python/utils/cloudsql.py]: fixed

File: src/main/python/utils/cloudsql.py (hunk 3 above)
Comment (TylerTollefson) [reply src/main/python/utils/cloudsql.py]: now passing config itself instead of making a new one

File: src/main/python/utils/cloudsql.py (hunk 4 above)
Comment (TylerTollefson) [reply src/main/python/utils/cloudsql.py]: this had been changed to insert_instance

File: src/main/wdl/cloudsql.wdl (hunk 5 above)
Comment (TylerTollefson) [reply src/main/wdl/cloudsql.wdl:20]: Added defaults to the inputs that are used for the API request

File: src/main/python/utils/cloudsql.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            print("done.")
+            operation_complete = True
+            if "error" in result:
+                raise Exception(result["error"])
+            return result
+
+        time.sleep(1)
+    return operation_complete
+
+def insert_instance(project_id, config):
```
Comment (TylerTollefson) [inline src/main/python/utils/cloudsql.py:39]: Need to output the instance object after it is created

File: src/main/wdl/cloudsql.wdl [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        requireSSL: requireSSL,
+        privateNetwork: privateNetwork
+    }
+
+    command {
+        csql ${"--project_id=" + projectId} ${"--credentials=" + credentials} insert ~{write_json(config)}
+    }
+
+    output {
+      File results = stdout()
```
Comment (TylerTollefson) [inline src/main/wdl/cloudsql.wdl:59]: Need to handle the output of the instance object, will need to add a struct for this

File: src/main/python/utils/cloudsql.py (hunk 6 above)
Comment (TylerTollefson) [reply src/main/python/utils/cloudsql.py:39]: done, now outputs the created instance

File: src/main/wdl/cloudsql.wdl (hunk 7 above)
Comment (TylerTollefson) [reply src/main/wdl/cloudsql.wdl:59]: this is now being handled

**Review by jmesterh — APPROVED** [review]

**Pull request merged**